src/main/java/com/yahoo/bullet/querying/partitioning/SimpleEqualityPartitioner.java

```diff
@@ -0,0 +1,215 @@
+     *
+     * @param config The non-null config containing settings for this class.
+     */
+    public SimpleEqualityPartitioner(BulletConfig config) {
+        delimiter = config.getAs(BulletConfig.EQUALITY_PARTITIONER_DELIMITER, String.class);
+        fields = (List<String>) config.getAs(BulletConfig.EQUALITY_PARTITIONER_FIELDS, List.class);
+        fieldSet = new HashSet<>(fields);
+        String defaultKey = Collections.nCopies(fields.size(), NO_FIELD).stream().collect(Collectors.joining(delimiter));
+        defaultKeys = Collections.singletonList(defaultKey);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This partitioner ensures that queries are not stored in duplicate by returning only key for a query (the list
+     * that is returned is of size 1).
+     *
+     * @param query The query to partition.
+     * @return The {@link List} containing the one key for this query.
+     */
+    @Override
+    public List<String> getKeys(Query query) {
+        Objects.requireNonNull(query);
+        List<Clause> filters = query.getFilters();
+        // If no filters or has non ANDs, default partition
+        if (filters == null || filters.isEmpty() || hasNonANDLogicals(filters)) {
+            return defaultKeys;
+        }
+
+        // For each unique field, get all the FilterClauses that define an operation on it.
+        Map<String, List<FilterClause>> fieldFilters = new HashMap<>();
+        filters.forEach(c -> this.mapFieldToFilters(c, fieldFilters));
+
+        // If not one equality filter per field and not one value per filter, default partition
+        if (fieldFilters.values().stream().anyMatch(this::hasInvalidFilterClauses)) {
+            return defaultKeys;
+        }
+
+        // Generate key in fields order and pad with NO_FIELD if no mapping present
+        String key = fields.stream().map(fieldFilters::get).map(this::getFilterValue).collect(Collectors.joining(delimiter));
+        // For the SimpleEqualityPartitioner, the query is mapped to exactly one key only.
+        return Collections.singletonList(key);
+    }
+
+    @Override
+    public List<String> getKeys(BulletRecord record) {
+        Map<String, String> values = getFieldValues(record);
+        // For fields not present (NO_FIELD mapped), this de-dupes them with the binary combination that ignores them
+        Set<String> keys = new HashSet<>();
+
+        int numberOfFields = fields.size();
+        int combinations = 1 << numberOfFields;
+        // Generate a truth table for all possible combinations of the fields when using the field value or not
+        for (int i = 0; i < combinations; i++) {
+            // This makes the paddedBinary have fields.size() chars where each one represents to include or not the field
+            String paddedBinary = StringUtils.leftPad(Integer.toBinaryString(i), numberOfFields, FALSE_CHAR);
+            keys.add(binaryToKey(paddedBinary, values));
+        }
```

Comment:
Simplify by IntStream.range(0, 1 << fileds.size()).map(...).collect(...) ?

Reply:
or a good old recursive solution...

recurse(i, keyParts, keys, values)
{
	if (i < fields.size())
	{
		keyParts[i] = NO_FIELD;
		recurse(i + 1, fields, keyParts, keys, values);
		keyParts[i] = values.get(fields.get(i));
		recurse(i + 1, fields, keyParts, keys, values);
	}
	else
		keys.add(Stream.of(keyParts).collect(Collectors.joining(delimiter));
}

recurse(0, new String[fields.size()], keys, values);

Reply:
Done